src/index.ts

```diff
@@ -0,0 +1,45 @@
+import { v4 as uuidv4 } from 'uuid';
+
+type AlgorithmFunction = () => string;
+
+const algorithms: Record<string, AlgorithmFunction> = {
+  uuid4: uuidv4,
+};
+
+function replacingAlgorithm(
+  memoizationTable: Record<string, string>,
+  _: unknown,
+  uriBase: string,
+  uuidAlgorithm: string,
+  originalUuid: string,
+) {
+  if (memoizationTable[`${uriBase}${originalUuid}`]) {
+    return memoizationTable[`${uriBase}${originalUuid}`];
+  } else {
+    const algorithm = algorithms[uuidAlgorithm] || uuidv4;
+    const finalUri = `"${uriBase}${algorithm()}"`;
+    memoizationTable[`${uriBase}${originalUuid}`] = finalUri;
+    return finalUri;
+  }
+}
+
+// We disable typescript linting here as the curry function should indeed accept any function and any arguments
+// eslint-disable-next-line  @typescript-eslint/no-explicit-any, @typescript-eslint/ban-types
+function currying(fn: Function, ...args: any[]) {
```

Comment:
It doesn't make sense to me to use a generic currying function that then leads to a weird function signature for the replacement function (by weird, I mean the `_: unknown` argument). To me it would make sense to just have the signature of the replacement function to be
`const replacementFunction = (memoizationTable: Record<string, string>) => (uriBase: string, ...etc) => {`.

Reply:
The unknown argument is not cause of the currying, that's cause of the replace function that sends this argument, the currying is because I wanted to have a replace function that you could send a memoization table in order to be easily tested in isolation if needed, rather than having the curried be the original version of it, as that's the cause of using the replace function

Reply:
What I meant was, if instead of having a `currying` function, which takes any arbitrary function and 'curries' the provided arguments, we could do this in a more specific way, which gives us more precise types. For example if we just have
```
type ReplacementFunction =
  (memoizationTable: Record<string, string>) =>
  (uriBase: string, uuidAlgorithm: string, originalUuid: string) =>
  string
```